File: python/routes/file.py
from flask import Blueprint, request, send_file

from middleware.auth import login_required
from response import ApiCode, fail, ok, success
from services import file_service

file_bp = Blueprint('file', __name__, url_prefix='/api/files')


def _parse_parent_id(raw):
    if raw is None or raw == '' or raw == 'null':
        return None
    try:
        return int(raw)
    except (TypeError, ValueError) as exc:
        raise ValueError('parentId 无效') from exc


@file_bp.get('')
@login_required
def list_files():
    try:
        parent_id = _parse_parent_id(request.args.get('parentId'))
    except ValueError as exc:
        return fail(str(exc), code=ApiCode.BAD_REQUEST)

    items = file_service.list_files(
        parent_id,
        keyword=request.args.get('keyword', ''),
        category=request.args.get('category', ''),
        ext=request.args.get('ext', ''),
        created_by=request.args.get('createdBy', ''),
    )
    breadcrumbs = file_service.get_breadcrumbs(parent_id)
    return ok({'list': items, 'breadcrumbs': breadcrumbs, 'parentId': parent_id})


@file_bp.get('/<int:file_id>')
@login_required
def get_file(file_id: int):
    item = file_service.get_file(file_id)
    if not item:
        return fail('文件不存在', code=ApiCode.NOT_FOUND, http_status=404)
    return ok(item)


@file_bp.post('/folder')
@login_required
def create_folder():
    data = request.get_json(silent=True) or {}
    try:
        parent_id = _parse_parent_id(data.get('parentId'))
        item = file_service.create_folder(parent_id, str(data.get('name', '')))
    except ValueError as exc:
        return fail(str(exc), code=ApiCode.BAD_REQUEST)
    return success(item, code=ApiCode.CREATE_SUCCESS, message='文件夹已创建', title='文件管理')


@file_bp.post('/upload')
@login_required
def upload_file():
    try:
        parent_id = _parse_parent_id(request.form.get('parentId'))
        upload = request.files.get('file')
        relative_path = str(request.form.get('relativePath') or '').strip()
        item = file_service.upload_file(parent_id, upload, relative_path=relative_path)
    except ValueError as exc:
        return fail(str(exc), code=ApiCode.BAD_REQUEST)
    return success(item, code=ApiCode.CREATE_SUCCESS, message='上传成功', title='文件管理')


@file_bp.put('/<int:file_id>/rename')
@login_required
def rename_file(file_id: int):
    data = request.get_json(silent=True) or {}
    try:
        item = file_service.rename_file(file_id, str(data.get('name', '')))
    except ValueError as exc:
        return fail(str(exc), code=ApiCode.BAD_REQUEST)
    return success(item, code=ApiCode.UPDATE_SUCCESS, message='已重命名', title='文件管理')


@file_bp.put('/<int:file_id>/move')
@login_required
def move_file(file_id: int):
    data = request.get_json(silent=True) or {}
    try:
        target = _parse_parent_id(data.get('parentId'))
        item = file_service.move_file(file_id, target)
    except ValueError as exc:
        return fail(str(exc), code=ApiCode.BAD_REQUEST)
    return success(item, code=ApiCode.UPDATE_SUCCESS, message='已移动', title='文件管理')


@file_bp.post('/delete')
@login_required
def delete_files():
    data = request.get_json(silent=True) or {}
    ids = data.get('ids') or []
    try:
        count = file_service.delete_files([int(i) for i in ids])
    except (TypeError, ValueError) as exc:
        return fail(str(exc) if str(exc) else '参数错误', code=ApiCode.BAD_REQUEST)
    return success(
        {'count': count},
        code=ApiCode.DELETE_SUCCESS,
        message=f'已删除 {count} 项',
        title='文件管理',
    )


@file_bp.get('/<int:file_id>/download')
@login_required
def download_file(file_id: int):
    try:
        path, name, mime = file_service.get_download_path(file_id)
    except ValueError as exc:
        return fail(str(exc), code=ApiCode.BAD_REQUEST)
    return send_file(path, mimetype=mime, as_attachment=True, download_name=name)


@file_bp.get('/<int:file_id>/raw')
@login_required
def raw_file(file_id: int):
    """预览用：图片/音视频/PDF 内联输出。"""
    try:
        path, name, mime = file_service.get_download_path(file_id)
    except ValueError as exc:
        return fail(str(exc), code=ApiCode.BAD_REQUEST)
    return send_file(path, mimetype=mime, as_attachment=False, download_name=name)


@file_bp.get('/<int:file_id>/text')
@login_required
def text_preview(file_id: int):
    try:
        data = file_service.read_text_preview(file_id)
    except ValueError as exc:
        return fail(str(exc), code=ApiCode.BAD_REQUEST)
    return ok(data)

File: src/api/file.ts
import { apiRequest } from './http'

export type FileCategory =
  | 'folder'
  | 'image'
  | 'video'
  | 'audio'
  | 'document'
  | 'office'
  | 'pdf'
  | 'markdown'
  | 'code'
  | 'archive'
  | 'install'
  | 'other'

export type FilePreviewKind =
  | 'folder'
  | 'image'
  | 'video'
  | 'audio'
  | 'pdf'
  | 'text'
  | 'office'
  | 'install'
  | 'download'

export interface FileEntry {
  id: number
  parentId: number | null
  name: string
  isFolder: boolean
  ext: string | null
  mimeType: string | null
  sizeBytes: number
  category: FileCategory
  createdBy: string | null
  updatedBy: string | null
  createdAt?: string
  updatedAt?: string
  previewKind: FilePreviewKind
  canPreview: boolean
}

export interface FileBreadcrumb {
  id: number | null
  name: string
}

export interface FileListResult {
  list: FileEntry[]
  breadcrumbs: FileBreadcrumb[]
  parentId: number | null
}

export interface FileTextPreview {
  name: string
  content: string
  truncated: boolean
}

export interface FileListQuery {
  parentId?: number | null
  keyword?: string
  category?: string
  ext?: string
  createdBy?: string
}

function authHeaders(): HeadersInit {
  const token = localStorage.getItem('maxadmin_token')
  return token ? { Authorization: `Bearer ${token}` } : {}
}

function buildQuery(query: FileListQuery = {}) {
  const params = new URLSearchParams()
  if (query.parentId != null) params.set('parentId', String(query.parentId))
  if (query.keyword) params.set('keyword', query.keyword)
  if (query.category) params.set('category', query.category)
  if (query.ext) params.set('ext', query.ext)
  if (query.createdBy) params.set('createdBy', query.createdBy)
  const qs = params.toString()
  return qs ? `?${qs}` : ''
}

export function fetchFiles(query: FileListQuery = {}) {
  return apiRequest<FileListResult>(`/api/files${buildQuery(query)}`)
}

export function fetchFile(id: number) {
  return apiRequest<FileEntry>(`/api/files/${id}`)
}

export function createFolder(parentId: number | null, name: string) {
  return apiRequest<FileEntry>('/api/files/folder', {
    method: 'POST',
    body: JSON.stringify({ parentId, name }),
  })
}

export function uploadFile(
  parentId: number | null,
  file: File,
  onProgress?: (percent: number) => void,
  relativePath?: string,
) {
  return new Promise<FileEntry>((resolve, reject) => {
    const xhr = new XMLHttpRequest()
    const form = new FormData()
    if (parentId != null) form.append('parentId', String(parentId))
    form.append('file', file)
    const rel = (relativePath || (file as File & { webkitRelativePath?: string }).webkitRelativePath || '').trim()
    if (rel) form.append('relativePath', rel)

    xhr.open('POST', '/api/files/upload')
    const token = localStorage.getItem('maxadmin_token')
    if (token) xhr.setRequestHeader('Authorization', `Bearer ${token}`)

    xhr.upload.onprogress = (e) => {
      if (e.lengthComputable && onProgress) {
        onProgress(Math.round((e.loaded / e.total) * 100))
      }
    }

    xhr.onload = () => {
      try {
        const payload = JSON.parse(xhr.responseText)
        if (payload.code === 0 || (payload.code >= 1000 && payload.code < 2000)) {
          resolve(payload.data as FileEntry)
        } else {
          reject(new Error(payload.message || '上传失败'))
        }
      } catch {
        reject(new Error('上传响应解析失败'))
      }
    }
    xhr.onerror = () => reject(new Error('网络异常'))
    xhr.send(form)
  })
}

export function renameFile(id: number, name: string) {
  return apiRequest<FileEntry>(`/api/files/${id}/rename`, {
    method: 'PUT',
    body: JSON.stringify({ name }),
  })
}

export function moveFile(id: number, parentId: number | null) {
  return apiRequest<FileEntry>(`/api/files/${id}/move`, {
    method: 'PUT',
    body: JSON.stringify({ parentId }),
  })
}

export function deleteFiles(ids: number[]) {
  return apiRequest<{ count: number }>('/api/files/delete', {
    method: 'POST',
    body: JSON.stringify({ ids }),
  })
}

export function fetchTextPreview(id: number) {
  return apiRequest<FileTextPreview>(`/api/files/${id}/text`)
}

export function fileRawUrl(id: number) {
  const token = localStorage.getItem('maxadmin_token')
  const q = token ? `?token=${encodeURIComponent(token)}` : ''
  // 浏览器 <img>/<video> 无法自定义 Authorization，提供带 token 查询的备用路径不方便改后端
  // 这里用 blob 下载方式；预览组件会用 fetch + blob URL
  return `/api/files/${id}/raw${q}`
}

export async function fetchFileBlob(id: number, asAttachment = false) {
  const path = asAttachment ? `/api/files/${id}/download` : `/api/files/${id}/raw`
  const res = await fetch(path, { headers: authHeaders() })
  if (!res.ok) throw new Error('文件加载失败')
  return res.blob()
}

export async function downloadFileById(id: number, filename: string) {
  const blob = await fetchFileBlob(id, true)
  const url = URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = url
  a.download = filename
  a.click()
  URL.revokeObjectURL(url)
}
